clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchDocumentMapper.scala

```diff
@@ -1,29 +1,52 @@
 package org.broadinstitute.clio.server.dataaccess.elasticsearch
 
+import io.circe.{Encoder, Json}
+import io.circe.syntax._
+import org.broadinstitute.clio.transfer.model.TransferMetadata
+import org.broadinstitute.clio.util.json.ModelAutoDerivation
+import org.broadinstitute.clio.util.model.UpsertId
+
 /**
   * Maps metadata to an Elasticsearch document.
   *
-  * @tparam ModelKey      The type of key.
-  * @tparam ModelMetadata The type of the metadata.
-  * @tparam Document      The type of the document.
+  * @tparam Key      The type of the TransferKey.
+  * @tparam Metadata The type of the TransferMetadata.
   */
-abstract class ElasticsearchDocumentMapper[ModelKey, ModelMetadata, Document] {
+class ElasticsearchDocumentMapper[Key <: Product: Encoder, Metadata <: TransferMetadata[
+  Metadata
+]: Encoder](genId: () => UpsertId)
+    extends ModelAutoDerivation {
 
   /**
-    * Returns an empty document for a key.
     *
-    * @param key The key.
-    * @return The newly initialized document.
+    * @param key      The TransferKey from which to create the document.
+    * @param metadata The TransferMetadata from which to create the document.
+    * @return
     */
-  def empty(key: ModelKey): Document
+  def document(key: Key, metadata: Metadata): Json = {
+    val upsertIdMap = Map(
+      ElasticsearchIndex.UpsertIdElasticsearchName -> genId()
+    )
+    val entityIdMap = Map(
+      ElasticsearchIndex.EntityIdElasticsearchName -> Symbol(
+        key.productIterator.mkString(".")
+      )
+    )
 
-  /**
-    * Returns an updated document with new metadata, where any fields that are `None` do NOT overwrite the existing
-    * metadata.
-    *
-    * @param document The existing document.
-    * @param metadata The new metadata.
-    * @return The updated document.
-    */
-  def withMetadata(document: Document, metadata: ModelMetadata): Document
+    val keyJson = key.asJson
+    val metadataJson = metadata.asJson
+    val upsertIdJson = upsertIdMap.asJson
+    val entityIdJson = entityIdMap.asJson
+    keyJson.deepMerge(metadataJson).deepMerge(upsertIdJson).deepMerge(entityIdJson)
+  }
+}
+
+object ElasticsearchDocumentMapper {
+
+  def apply[Key <: Product: Encoder, Metadata <: TransferMetadata[Metadata]: Encoder]: ElasticsearchDocumentMapper[
```

Comment:
Relaxing the above constraint will let this be less insane too.

Reply:
I think I do need the `<: Product` in order for `productIterator` to work on line 32